crates/bevy_ecs/src/schedule/stage.rs

```diff
@@ -402,33 +345,34 @@ impl SystemStage {
             })
             .collect();
 
-        for index in self.uninitialized_at_start.drain(..) {
-            let container = &mut self.exclusive_at_start[index];
-            if let Some(index) = container.run_criteria() {
-                container.set_run_criteria(new_indices[index]);
-            }
+        // initialize the systems and sort them into the appropriate vectors
+        for (mut container, system_type) in self.uninitialized_systems.drain(..) {
             container.system_mut().initialize(world);
-        }
-        for index in self.uninitialized_before_commands.drain(..) {
-            let container = &mut self.exclusive_before_commands[index];
-            if let Some(index) = container.run_criteria() {
-                container.set_run_criteria(new_indices[index]);
-            }
-            container.system_mut().initialize(world);
-        }
-        for index in self.uninitialized_at_end.drain(..) {
-            let container = &mut self.exclusive_at_end[index];
+
             if let Some(index) = container.run_criteria() {
                 container.set_run_criteria(new_indices[index]);
             }
-            container.system_mut().initialize(world);
-        }
-        for index in self.uninitialized_parallel.drain(..) {
-            let container = &mut self.parallel[index];
-            if let Some(index) = container.run_criteria() {
-                container.set_run_criteria(new_indices[index]);
+
+            match system_type {
+                SystemType::Parallel => {
```

Comment:
The `SystemType` enum s variants should probably be renamed to be more descriptive given that now "parallel" systems can actually be exclusive.

Reply:
Renamed to what?

There's exclusive in the "this system has an `&mut World` param" sense and in the "this system is scheduled with one of the 3 insertion points" sense, so I think it's fine.

I can fix exclusive systems being "mistyped" as parallel tho (this is corrected in `SystemStage`), since I realize it's now possible to detect which systems are immediately exclusive with `is_exclusive`.

Reply:
That'd work. I mostly don't want exclusive systems under the parallel umbrella is all.

Reply:
Should be good now. Exclusive systems are tagged with `SystemType::Exclusive` when the descriptor is constructed.